nni/mutable/utils.py

```diff
@@ -0,0 +1,226 @@
+
+        # for example, key: "param"
+        self.key = key
+
+        # the "path" of current name
+        # By default, it's ``[]``
+        # If a LabelNamespace is nested inside another one, it will become something like ``[1, 3, 2]``.
+        # See ``__enter__``.
+        self.name_path: list[int] = []
+
+    def __enter__(self):
+        # Enter the label namespace resets the counter associated with the namespace.
+        #
+        # It also pushes itself into the stack, so as to support nested namespace.
+        # For example, currently the top of stack is [1, 2, 2], and [1, 2, 2, 3] is used,
+        # the next thing up is [1, 2, 2, 4].
+        # `reset_uid` to count from zero for "param_1_2_2_4"
+        try:
+            parent_context: LabelNamespace = LabelNamespace.current(self.key)
+            next_uid = uid(parent_context._simple_name())
+            self.name_path = parent_context.name_path + [next_uid]
+            ContextStack.push(self.key, self)
+            reset_uid(self._simple_name())
+        except NoContextError:
+            # not found, no existing namespace
+            self.name_path = []
+            ContextStack.push(self.key, self)
+            reset_uid(self._simple_name())
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        ContextStack.pop(self.key)
+
+    def _simple_name(self) -> str:
+        return self.key + ''.join(['_' + str(k) for k in self.name_path])
+
+    def __repr__(self):
+        return f'LabelNamespace(name={self._simple_name()})'
+
+    @staticmethod
+    def current(key: str = _DEFAULT_LABEL_NAMESPACE) -> LabelNamespace:
+        """Get the current namespace associated with the key.
+
+        Examples
+        --------
+        >>> with LabelNamespace("param") as namespace:
+        ...     # somewhere in the middle of the code.
+        ...     LabelNamespace.current("param")     # Will return the same namespace
+
+        Raises
+        ------
+        NoContextError
+            If no context is found.
+        """
+        return ContextStack.top(key)
+
+
+def auto_label(key: str = _DEFAULT_LABEL_NAMESPACE) -> str:
+    """Automatically generate a label in case the label is not set.
+
+    It will look for the nearest :class:`LabelNamespace` with the specified key.
+    If not found, it prints a warning and use the ``global`` namespace.
+
+    Parameters
+    ----------
+    key
+        The key of the namespace to use. If not specified, the default namespace (param) will be used.
+
+    Examples
+    --------
+    >>> label1 = auto_label()               # global_1
+    >>> with LabelNamespace('param'):
+    ...     label2 = auto_label('param')    # param_1, because in the namespace "param"
+    >>> with LabelNamespace():
+    ...     label3 = auto_label()           # param_1, default key is used
+    >>> with LabelNamespace('another'):
+    ...     label4 = auto_label('another')  # another_1
```

Comment:
I don't understand why under `'another'` context, I should pass `'another'` to `auto_label`

Reply:
Suppose you have:

```python
with LabelNamespace('abc'):
  with LabelNamespace('def'):
     auto_label()  # abc_1 or def_1?
```

Maybe I should auto-fetch the nearest one?

Reply:
I don't know the usage scenario very well, my intuition in this case is `abc_def_1`.

Reply:
In current implementation:

This will generate an extra level.

```python
with LabelNamespace('abc'):
  with LabelNamespace('abc'):
    auto_label('abc')  # abc_1_1
    auto_label()  # global_1
```

The following is actually identical to `with LabelNamespace('abc'), LabelNamespace('def'):`. The nested-with here has no effect. 

```python
with LabelNamespace('abc'):
  with LabelNamespace('def'):
    auto_label('abc')  # abc_1
    auto_label('def')  # def_1
    auto_label()  # global_1
```

I think `auto_label()` can automatically fetch the nearest label namespace if you agree it's a good idea.

Nested-with can also be concatenated to generate labels like `abc_abc_1` or `abc_def_1`. Or maybe `abc/abc/1` and `abc/def/1`?

Reply:
agree, it seems more auto if it can fetch the nearest label namespace.

I have no idea if we should support the nested labels generation, if we do, I prefer `abc/abc/1`.

Reply:
If then, I need to rewrite the whole label namespace. :)